ow_power_system/src/PowerSystemNode.cpp

```diff
@@ -71,154 +52,14 @@ bool PowerSystemNode::loadSystemConfig()
-  }
-  return result;
-}
-
-bool PowerSystemNode::loadCustomFaultPowerProfile(string path, string custom_file)
-{
-  m_custom_power_fault_sequence = loadPowerProfile(path, custom_file);
-
-  // Return false if the sequence was not properly initialized.
-  return (m_custom_power_fault_sequence.size() > 0);
-}
-
-bool PowerSystemNode::initPrognoser()
-{
-  // Create a configuration from a file
-  auto prognoser_config_path = ros::package::getPath("ow_power_system") + "/config/prognoser.cfg";
-  ConfigMap prognoser_config(prognoser_config_path);
-
-  // Contruct a new prognoser using the prognoser factory. The prognoser
-  // will automatically construct an appropriate model, observer and predictor
-  // based on the values specified in the config.
-  m_prognoser = PrognoserFactory::instance().Create("ModelBasedPrognoser", prognoser_config);
-
-  if (m_prognoser == nullptr)
-  {
-    return false;
-  }
-
-  // Initialize the GSAP prognoser
-  auto init_data = composePrognoserData(m_initial_power, m_initial_voltage, m_initial_temperature);
-  m_prognoser->step(init_data);
-
-  m_init_time = system_clock::now();
+  m_time_interval = system_config.getInt32("time_interval");
 
   return true;
 }
 
-bool PowerSystemNode::initTopics()
+void PowerSystemNode::applyMechanicalPower(double mechanical_power)
 {
-  // Construct the PowerSystemNode publishers
-  m_mechanical_power_raw_pub = m_nh.advertise<Float64>("mechanical_power/raw", 1);
-  m_mechanical_power_avg_pub = m_nh.advertise<Float64>("mechanical_power/average", 1);
-  m_battery_state_of_charge_pub = m_nh.advertise<owl_msgs::BatteryStateOfCharge>("/battery_state_of_charge", 1);
-  m_battery_remaining_useful_life_pub = m_nh.advertise<owl_msgs::BatteryRemainingUsefulLife>("/battery_remaining_useful_life", 1);
-  m_battery_temperature_pub = m_nh.advertise<owl_msgs::BatteryTemperature>("/battery_temperature", 1);
-  // Finally subscribe to the joint_states to estimate the mechanical power
-  m_joint_states_sub = m_nh.subscribe("/joint_states", 1, &PowerSystemNode::jointStatesCb, this);
-  return true;
-}
-
-void PowerSystemNode::jointStatesCb(const sensor_msgs::JointStateConstPtr& msg)
-{
-  auto power_watts = 0.0;  // This includes the arm + antenna
-  for (auto i = 0; i < ow_lander::NUM_JOINTS; ++i)
-    power_watts += fabs(msg->velocity[i] * msg->effort[i]);
-
-  m_power_values[++m_power_values_index % m_power_values.size()] = power_watts;   // [W]
-  auto mean_mechanical_power =
-      accumulate(begin(m_power_values), end(m_power_values), 0.0) / m_power_values.size();
-
-  Float64 mechanical_power_raw_msg, mechanical_power_avg_msg;
-  mechanical_power_raw_msg.data = power_watts;
-  mechanical_power_avg_msg.data = mean_mechanical_power;
-  m_mechanical_power_raw_pub.publish(mechanical_power_raw_msg);
-  m_mechanical_power_avg_pub.publish(mechanical_power_avg_msg);
-
-  m_unprocessed_mechanical_power = mean_mechanical_power;
+  m_unprocessed_mechanical_power = mechanical_power;
```

Comment:
The naming convention (and comment a few lines below) imply accumulation, but it's been a value assignment for as long as I can remember. I'll do some tests and see if changing it to ``+=`` affects any behavior, though I don't think it will because I remember checking this block of code some time ago and finding that the ``if`` statement always ran, so it never had to accumulate in the first place.

Essentially, the ``if`` statement was there to withhold mechanical power from being updated if the loop was currently processing the previous mechanical power value in ``runPrognoser`` (since it's a callback function it could potentially interrupt). Because the assignment allows for one interrupt with no issue, there would only be an observable problem of losing unprocessed power if two callback interrupts occurred before ``runPrognoser`` finished once. This never happens because of how long it takes for GSAP to return predictions.

That said, I'll probably still change this to a ``+=`` since at worst it won't affect anything and it will make more sense with the naming of the variables and the comments.

Reply:
I won't take the time to look at it in depth right now, but I think it will be worth investigating these values and how they work in that ticket about the mechanical power moving average window. For now, if the if statement is always hit, I think your change is fine, but there is something about how this moving average works that bothers me. I may also just be misunderstanding something.